Updates to the ProblemChild and DGA integration packages

ajosh0504 opened the pull request:
## What does this PR do?
- Fixes some bugs in/arising from the ProblemChild and DGA integrations packages.
- Some changes to improve discoverability of rules associated with the two packages.

## Changes made:
- ProblemChild integration package
    - Fix a script that was causing the inference pipeline to fail
    - Added additional tags to rules associated with the integration package for easy discoverability
    - Updated the query and influencers in `problemchild-ml.json`
    - Updated all rule versions to 2 since this is an actual update to the rules (there were some rules which started with a version 2, but we can probably disregard that since this is still an experimental package) 
- DGA integration package
    - Fixed a rule that was pointing to a non-existent ML job
    -  Added additional tags to rules associated with the integration package for easy discoverability
    - Corrected the n-gram feature creation in `dga_inference_pipeline` script.
    - Updated all rule versions to 2 since this is an actual update to the rules (there were some rules which started with a version 2, but we can probably disregard that since this is still an experimental package)

## Checklist
- [x] I have reviewed [tips for building integrations](https://github.com/elastic/integrations/blob/main/docs/tips_for_building_integrations.md) and this pull request is aligned with them.
- [x] I have verified that all data streams collect metrics or logs.
- [x] I have added an entry to my package's `changelog.yml` file.
- [x] I have verified that Kibana version constraints are current according to [guidelines](https://github.com/elastic/elastic-package/blob/master/docs/howto/stack_version_support.md#when-to-update-the-condition).

## Related issues
- Closes #3723 
- Closes #3738
- Closes #3737
- Closes #3762

elasticmachine commented:
## :broken_heart: Build Failed


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-3777/detail/PR-3777/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-3777/detail/PR-3777/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-3777/detail/PR-3777/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://beats-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-3777/detail/PR-3777/1//artifacts) [![preview](https://img.shields.io/badge/docs-preview-yellowgreen)](http://integrations_3777.docs-preview.app.elstc.co/diff)  [![preview](https://img.shields.io/badge/elastic-observability-blue)](https://ci-stats.elastic.co/app/apm/services/beats-ci/transactions/view?rangeFrom=2022-07-20T21:48:14.469Z&rangeTo=2022-07-20T22:08:14.469Z&transactionName=Ingest-manager/integrations/PR-{number}&transactionType=job&latencyAggregationType=avg&traceId=abd9122795f777da996023d2bca968bc&transactionId=6884386c9951bcf8)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats




* Start Time: 2022-07-20T21:58:14.469+0000


* Duration: 12 min 3 sec





</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY -->














### Steps errors [![3](https://img.shields.io/badge/3%20-red)](https://beats-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-3777/detail/PR-3777/1//pipeline)
  <details><summary>Expand to view the steps failures</summary>
  <p>


  
  
  
  
##### `Check integration: dga`
<ul>
<li>Took 0 min 0 sec . View more details <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Ingest-manager/pipelines/integrations/pipelines/PR-3777/runs/1/steps/910/log/?start=0">here</a></li>
<li>Description: <code>../../build/elastic-package check -v</code></l1>
</ul>
  
  
  
  
##### `Check integration: problemchild`
<ul>
<li>Took 0 min 0 sec . View more details <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Ingest-manager/pipelines/integrations/pipelines/PR-3777/runs/1/steps/894/log/?start=0">here</a></li>
<li>Description: <code>../../build/elastic-package check -v</code></l1>
</ul>
  
  
  
  
##### `Google Storage Download`
<ul>
<li>Took 0 min 0 sec . View more details <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Ingest-manager/pipelines/integrations/pipelines/PR-3777/runs/1/steps/959/log/?start=0">here</a></li>

</ul>
  
  </p>
  </details>



## :robot: GitHub comments

To re-run your PR in the CI, just comment with:



- `/test` : Re-trigger the build.




<!--COMMENT_GENERATED_WITH_ID_comment.id-->

ajosh0504 commented:
@mtojek Does this update to the packages count as a new version or the same? I was hoping for this to be a sort of bug fix PR for the existing package. 

P.S. I just noticed that some of the rules in the initial packages have some version 2s- maybe we want to reset those to 1? I can take care of that in this PR too.

mtojek reviewed (commented):
> Please check to make sure the structure of the data_stream directory in both the packages looks correct

Please make sure that CI passes for this PR. CI uses elastic-package tool to validate against the package-spec.

> Is a manifest.yml file in the [data_stream/predictions](https://github.com/elastic/integrations/tree/problemchild_dga_update/packages/problemchild/data_stream/predictions) directory mandatory? 

Manifest describes the data stream and if you want to have data stream it's mandatory. I see that you don't have much experience with package development, so you may find this [blog post](https://www.elastic.co/blog/use-elastic-package-to-create-and-bootstrap-a-new-elastic-integration) helpful.

> What should it contain?

As you replied to you over Slack, the [content](https://github.com/elastic/package-spec/blob/main/versions/1/integration/data_stream/manifest.spec.yml) is defined by package-spec. Any property that is "legal", must be defined by spec. 

> Since I am adding new index mappings to the packages, is there anything else required other than this integrations PR?

I don't know, not sure what are you trying to achieve. You've implemented an incomplete data stream - just added a few fields, and modified ingest pipeline and security rules. Please sync with @alvarezmelissa87 who has already successfully contributed to Integrations.

> Does this update to the packages count as a new version or the same? I was hoping for this to be a sort of bug fix PR for the existing packages.

If you want to release those changes, you need to bump up the version and add a corresponding changelog entry. There is a background job set to determine if the package version has been updated and if so, it syncs changes with Package Storage.

ajosh0504 commented:
@mtojek Thanks for the detailed response. Looks like CI is indeed failing on the missing manifests now.

I was reading your blog post but looks like the example you walkthrough is for a new integration, creating a new data stream. I'm working with an existing package and existing data streams, specifically the `logs-*` , `winlogbeat-*` and `endgame-*` data streams, based on which indices the user wants to run the ingest pipeline (defined in the top-level `elasticsearch folder) on.  It seems to me then that I should have the ability to define an index template asset in the top-level folder as well that doesn't seem to be supported by [package-spec](https://github.com/elastic/package-spec/blob/main/versions/1/integration/elasticsearch/spec.yml).

Eitherway, I'll follow-up with @alvarezmelissa87 and see where we end up on this. Thanks!

ajosh0504 commented:
Tested the packages with `elastic-package` to make sure that the bug fixes worked.

andrewkroh commented on packages/dga/changelog.yml:
```diff
@@ -1,3 +1,8 @@
+- version: "0.0.3"
+  changes:
+    - description: Fix some bugs in the DGA package
```
As a user I prefer some specific details about bugs that were fixed. This will render in our docs at https://docs.elastic.co/en/integrations/dga#changelog.

andrewkroh commented on packages/problemchild/elasticsearch/ingest_pipeline/problem_child_inference_pipeline.yml:
```diff
@@ -78,18 +78,14 @@ processors:
-        if (ctx.containsKey('process') && ctx['process'].containsKey('Ext') && ctx['process']['Ext'].containsKey('code_signature') && ctx['process']['Ext']['code_signature'].containsKey('exists')) {
-          ctx.feature_process_signed = ctx['process']['Ext']['code_signature']['exists']
-        } else if (ctx.containsKey('process') && ctx['process'].containsKey('code_signature') && ctx['process']['code_signature'].containsKey('exists')) {
+        if (ctx.containsKey('process') && ctx['process'].containsKey('code_signature') && ctx['process']['code_signature'].containsKey('exists')) {
```
This can safely be expressed as `ctx.process?.code_signature?.exists != null` using the [null-safe operator](https://www.elastic.co/guide/en/elasticsearch/painless/current/painless-operators-reference.html#null-safe-operator).

ajosh0504 replied:
We could cover this in a follow-up PR, since this PR only deals with bug fixes?

alvarezmelissa87 commented on packages/problemchild/changelog.yml:
```diff
@@ -1,3 +1,8 @@
+- version: "0.0.3"
+  changes:
+    - description: Fix some bugs in the ProblemChild package
```
Probably want to be more specific about what was fixed here, as well.

alvarezmelissa87 commented on packages/problemchild/kibana/ml_module/problemchild-ml.json:
```diff
@@ -9,35 +9,19 @@
+                "minimum_should_match": 1,
                 "should": [
                     {
-                        "bool": {
```
Could you explain why this filter isn't needed anymore?

ajosh0504 replied:
The original query basically translated to checking whether the fields ["process.name", "host.hostname", "process.command_line"] were present OR if the `problemchild.prediction_probability` field was present.

This is not what we want to check for, I don't think. We want to check if the events being passed to the anomaly detection jobs either have a `problemchild. prediction` or a `blocklist_label`, which is what I changed the query to.

As an aside, I also cross-checked with DGA, and the query there seems to match what's in the datafeeds as well.

ajosh0504 replied:
Is that not what the query should reflect?

alvarezmelissa87 commented:
Overall LGTM. Re: the questions in the description - I don't see any `data_stream` changes in this PR so maybe these comments are out of date?

ajosh0504 commented:
> Overall LGTM. Re: the questions in the description - I don't see any `data_stream` changes in this PR so maybe these comments are out of date?

Right, updated the description.

ajosh0504 commented:
@alvarezmelissa87 Here's screenshots to prove that things are working as expected:

**DGA**
- Making sure that the right version is being used for testing:
<img width="1282" alt="Screen Shot 2022-08-02 at 3 18 04 PM" src="https://user-images.githubusercontent.com/30438249/182483080-cfb67f80-eff7-465d-b945-94a18ff08dd2.png">

- Removed reference to a non-existent rule:
<img width="1643" alt="Screen Shot 2022-08-02 at 3 14 32 PM" src="https://user-images.githubusercontent.com/30438249/182482746-8028d673-70a8-4af7-89f7-fe1de89f7f80.png">

- `DGA` tag allows filtering on all DGA-related rules
<img width="1643" alt="Screen Shot 2022-08-02 at 3 14 05 PM" src="https://user-images.githubusercontent.com/30438249/182482872-98772c0b-9ee8-491c-b2a2-311a27238c93.png">

**ProblemChild**
- Predictions were previously not coming through due to a bug in the inference pipeline. Seeing predictions on Windows process events now:
<img width="1899" alt="Screen Shot 2022-08-02 at 3 11 11 PM" src="https://user-images.githubusercontent.com/30438249/182483232-a2c98a19-36ec-4876-b3a4-b1832e81b9c9.png">

- Anomaly Detection job datafeed was previously failing to start due to mapping errors. Added mappings to the component template (confirmed with @andrewkroh that this is the way to add custom mappings to existing datafeeds) and not seeing that error anymore:
<img width="493" alt="Screen Shot 2022-08-02 at 3 21 11 PM" src="https://user-images.githubusercontent.com/30438249/182483495-83ff95cc-d534-4bc0-987a-304c0edc39ef.png">
<img width="1625" alt="Screen Shot 2022-08-02 at 3 15 05 PM" src="https://user-images.githubusercontent.com/30438249/182483556-423ca6ec-5fa2-41e8-98b0-f7e159938f64.png">

- `LotL` tag allows filtering for rules related to the LotL Attack Detection package:
<img width="1648" alt="Screen Shot 2022-08-02 at 3 11 55 PM" src="https://user-images.githubusercontent.com/30438249/182483732-f607b20f-86a9-4f01-95a4-0209f9b242cd.png">

alvarezmelissa87 reviewed (approved):
LGTM ⚡

szeitlin commented on packages/dga/kibana/security_rule/64116bb2-0f2c-4cf6-9df4-9973452b4d4b.json:
```diff
@@ -14,8 +14,13 @@
+            "DGA",
+            "ML",
+            "Supervised",
+            "Anomalies"
```
how is Anomalies meant to be used? Seems kinda generic?

ajosh0504 replied:
We want to expose Anomalies vs Advanced Threats in the centralized Entity Analytics page. Anomalies would be things flagged as suspicious by individual ML jobs. Advanced Threats are things flagged as suspicious by more than one method. 

For example, in ProblemChild, suspicious events are identified by a supervised model and an anomaly detection job.

ML is a generic tag for all ML jobs in the Security solution. Supervised vs Unsupervised is based on which model contributed to the event being flagged as anomalous. Now that I think about it, this might be too much information for the user.

szeitlin replied:
This helps a lot, thank you! 

I think maybe it would help to write this up in a separate doc (since it might be hard to track back to this ticket later)

szeitlin replied:
ok, discussed with Sourin and put a couple comments in the spreadsheet, but the short version is that as far as the product is concerned, we only need 3 kinds of tags:

1) package tags - what rules and dashboards are tied to a package
2) threat category (internal vs. external threat) for future product features
3) entity tag (user vs. host) 

For the others, if we want internal tags for our own tracking/packaging/debugging/support purposes, that's good to know, just want to make sure we're keeping an eye on that as we along, so we don't end up with a proliferation of unused/overlapping/incorrect tag tech debt like we have on the malware pipelines.

szeitlin commented on packages/dga/kibana/security_rule/997ec71d-bddc-4513-b6f1-193f601fd420.json:
```diff
@@ -21,10 +21,14 @@
             "ML",
-            "DGA"
+            "Unsupervised",
+            "Advanced Threats"
```
what constitutes an Advanced Threat? Do we have a list of what's included in those? (also, why am I having deja vu as I write this?)

ajosh0504 replied:
Refer to the comment above.
